### PR TITLE
feat: port replay to HogQL filters (without changing UI/API 🤞)

### PR DESCRIPTION
**Pull request by daibhin**
towards https://github.com/PostHog/posthog/issues/22054

these change are behind a flag so we're safe to merge these (pretty much) whenever they're green and we can start checking performance in prod

### CODE REVIEW

File: posthog/hogql/property.py
Comment (pauldambra) [inline posthog/hogql/property.py:1]: @PostHog/team-product-analytics we're porting the replay filters to HogQL

this is the only change outside of of replay only things so far, does it seem safe?

File: posthog/hogql/database/schema/session_replay_events.py [hunk 1]
```diff
@@ -11,11 +11,70 @@
...
+    # because the time in the query will be different from the time in the test
+    # so we generate now in Python and pass it in
+    now = datetime.now()
+    clamp_to_ttl = ast.CompareOperation(
+        op=ast.CompareOperationOp.GtEq,
+        left=ast.Field(chain=chain),
+        right=ast.ArithmeticOperation(
+            op=ast.ArithmeticOperationOp.Sub,
+            left=ast.Constant(value=now),
+            # TODO be more clever about this date clamping
```
Comment (marandaneto) [inline posthog/hogql/database/schema/session_replay_events.py:73]: do we intend to fix this before shipping?

File: posthog/hogql/database/schema/test/test_session_replay_events.py [hunk 2]
```diff
@@ -1,17 +1,215 @@
...
+        response = execute_hogql_query(
+            parse_select(
+                "select session_id, any(person.properties.person_property) from raw_session_replay_events group by session_id order by session_id asc",
+            ),
+            self.team,
+        )
+
+        assert response.results == [
+            ("session_for_person_p1", None),
+            # todo: why is this the string true?
```
Comment (marandaneto) [inline posthog/hogql/database/schema/test/test_session_replay_events.py:201]: do we intend to investigate this before shipping?

File: posthog/session_recordings/queries/session_recording_list_from_filters.py [hunk 3]
```diff
@@ -9,14 +20,242 @@ class SessionRecordingQueryResult(NamedTuple):
...
+
+        for entity in self._filter.entities:
+            if entity.type == TREND_FILTER_TYPE_ACTIONS:
+                action = entity.get_action()
+                event_names.update([ae for ae in action.get_step_events() if ae not in event_names])
+            else:
+                if entity.id and entity.id not in event_names:
+                    event_names.add(entity.id)
+
+            # TODO: we're not passing the "right" type in here - should we change the signature or do something else?
```
Comment (marandaneto) [inline posthog/session_recordings/queries/session_recording_list_from_filters.py:104]: do we intend to fix this before shipping?

File: posthog/session_recordings/queries/session_recording_list_from_filters.py [hunk 4]
```diff
@@ -9,14 +20,242 @@ class SessionRecordingQueryResult(NamedTuple):
...
+                )
+            )
+
+        (event_where_exprs, _) = self._event_predicates
+        if event_where_exprs:
+            exprs.append(ast.Or(exprs=event_where_exprs))
+
+        if self._filter.property_groups:
+            # TRICKY: for person properties the scope of replay is equivalent to scope event, the session_replay_events schema mirrors events for person joining
+            # TODO: need to check multiple property types from replay queries
```
Comment (marandaneto) [inline posthog/session_recordings/queries/session_recording_list_from_filters.py:177]: should we do something about this?

**Review by marandaneto — APPROVED** [review]
left a few comments asking about the TODOs, it's unclear if they should be fixed before merging or if it's okay to be fixed later, just double-checking, otherwise LGTM (trusting you both since I don't know much about this piece of code =D).

**Comment by pauldambra** [comment]
yep, all the TODOs can stay for now... we've added a switch in the API call so you only run this code if you send a particular query parameter, which is only sent if a flag is enabled in the front end...

it's all a little too big to do in one step but don't want to lose sight of the TODOs as we go 😊

**Pull request merged**